docs/proposals/nodesets.md

```diff
@@ -0,0 +1,157 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/nodesets.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# NodeSets in Kubernetes
+
+**Author**: Justin Santa Barbara
+
+**Status**: Proposal.
+
+This document presents the design of the Kubernetes NodeSet, describes use cases, and gives an overview of the code.
+
+## Motivation
+
+The official bring-up procedure for a kubernetes cluster is to use kube-up.  This is a pretty complicated bash script
+that is largely separate for each provider.  It offers limited functionality: you cannot change the number of nodes
+directly, nor change the instance type used.  Many people resort to alternative methods of installation,
+to varying degrees of success.
+
+NodeSets put node launching under the control of the master.  They allow the operator to create
+and destroy NodeSets dynamically, or to change the number of instances in a NodeSet.  They
+allow for easier upgrades by launching nodes with different OS/k8s versions.
+
+This simplifies the Kubernetes installation, because the primary requirement is simply to launch a master.  Then
+the installation script can either automatically create a NodeSet, or leave this up to the operator post-creation.
+
+## Use Cases
+
+### Resize number of Nodes through the API
+
+By updating the count on a NodeSet, nodes can be launched or destroyed.  Users will no longer have
+to go to the underlying cloud API.  Kubernetes can likely make better decisions about which instance to
+terminate than a cloud-implemented auto-scaling group can, because it is more aware of the workloads.
+
+### Consistent auto-scaling across providers
+
+In future we can add auto-scaling functionality at the Kubernetes level (instead of at the cloud level), so
+that it works consistently across clouds, and can be based on Kubernetes metrics instead of cloud-metrics
+(i.e. it can take account of pods _not_ scheduled).
+
+### Support multiple NodeSets of different instance types
+
+Users may want to run instances of different types in the same cluster, permanently or transiently.
+Permanently mixing instance types may require more scheduler support, but for transient usage (changing
+instance sizes) this is helpful for changing a cluster.
+
+Users may also want to mix instance types in a way that is not visible to the cluster, for example mixing
```

Comment:
We will eventually want to surface such information to Kubernetes, since it would affect the SLOs of pods scheduled on those nodes.